Added functions to draw basic shapes

tpk32 opened the pull request:
** closes issue** #6 

Added the following
-  `drawLine()`  : draws a line from one point to another
-  `drawRectangle()` :  draws a rectangle with given endpoints of a diagonal
-  `drawCircle()` : draws a circle with a given center and radius
-  `getPixelColor()` : returns the color of a pixel as MinGLColor variable
-  updated MinGLColor struct to compare colors using `==` and `!=` operators

tpk32 committed:
Added line, rectangle, circle drawing function, getting pixel color and comparing pixel colors

tpk32 committed:
updated readme

tpk32 committed:
updated readme

amengol commented on MinGL.cpp:
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
 }
 
+const MinGLColor MinGL::getPixelColor(int x, int y, int width, int height) const{
+	int pixelSize = 4; // RGBA has 4 components (R, G, B, A)
```
Please use constexpr whenever possible, like here

tpk32 replied:
used constexpr for `pixelsize`  and static constexpr for `epsilon` in MinGLColor struct

amengol commented on MinGL.cpp:
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
 
+const MinGLColor MinGL::getPixelColor(int x, int y, int width, int height) const{
+	int pixelSize = 4; // RGBA has 4 components (R, G, B, A)
+    unsigned char* pixel = new unsigned char[pixelSize];
```
This is clearly leaking. Clean up or better yet, use a vector.

tpk32 replied:
changed to vector

amengol commented on MinGL.cpp:
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
 	glDisable(GL_SCISSOR_TEST);
 }
 
+const MinGLColor MinGL::getPixelColor(int x, int y, int width, int height) const{
```
Not used in this function. Please remove it.

tpk32 replied:
thanks for the review, I've removed `width` and `height` parameters, also removed the redundant `#include<queue>`

amengol commented on MinGL.h:
```diff
@@ -1,10 +1,25 @@
+		}
+		return true;
+	}
+	bool operator != (const MinGLColor color2) const{
+		for(int i=0; i<4; i++){
+			if( fabs(this->rgba[i] - color2.rgba[i]) > epsilon ) return true;
+		}
+		return false;
+	}
```
Simplify to 
`return !(*this == color2);`

tpk32 replied:
done

amengol commented on MinGL.cpp:
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
+	// specify center (x,y) of circle and its radius
+	// Bresenham's circle drawing algorithm is used here
+
+	auto draw = [&](int x, int y){
```
No need to capture everything by reference here, since most are just integers. If you must, capture only color by reference.

tpk32 replied:
captured only color by reference

amengol reviewed (commented):
Thanks for contributing! :)

amengol reviewed (changes requested):
Please address review comments.

tpk32 committed:
updated MinGLColor struct != operator, made pixelsize constexpr in getPixelColor()

tpk32 committed:
made some variables const, functions declared const, updated capture clause in drawCircle()

tpk32 committed:
Corrected C++ version required in readme.md

tpk32 commented:
Changes made:
- `epsilon` in MinGLColor struct is declared as `static constexpr`
- `getPixelColor()`  returns `MinGLColor` instead of `const MinGLColor` 
- `drawLine()`, `drawRectangle()` and `drawCircle()` functions are now declared `const` 
- `pixelSize` in `getPixelColor()` is declared `constexpr`
- used `vector`  in `getPixelColor()` instead of allocating memory for char array
- `dx` and `dy` in `drawLine()` are now const
- `xa`, `ya`, `xb`, `yb` in `drawRectangle()` are now const
- `draw` lambda in `drawCircle()` captures `color` by reference and `xc`,  `yc`, and `this` pointer by value

tpk32 commented:
Please review the changes and let me know if any other adjustments are required.

amengol commented on Example/shapesExample.cpp:
```diff
@@ -0,0 +1,48 @@
+	if (!minGL.init(width, height, "MinGL"))
+		return -1;
+
+	const char* glsl_version = "#version 130";
```
Not sure why this glsl_version line is here!

tpk32 replied:
this line was included in [Main.cpp](https://github.com/tpk32/MinGL/blob/main/Example/Main.cpp) and i copied some code from there, that's why its here also. I don't know if it serves some purpose but the code runs fine without it.
Shall i remove this line from both [Main.cpp](https://github.com/tpk32/MinGL/blob/feature-MinGL-adding-basic-shapes/Example/Main.cpp) and [shapesExample.cpp](https://github.com/tpk32/MinGL/blob/feature-MinGL-adding-basic-shapes/Example/shapesExample.cpp)

amengol replied:
Hmmm. I see now. Yeah, it was probably a leftover from something I was trying to do at the time.
Can you please remove from both?

Now that I see it, I don't like that it has two main() functions. We're not JAVA! :)
I might be asking too much, but can you try moving the examples to a Header file with functions like drawCube and DrawShapes and call them from "Main.cpp"? You can leave one of the examples commented out, no problem.
Thanks a lot!

amengol reviewed (changes requested):
Looking good. Just address the glsl version comment. Thanks!

tpk32 committed:
Moved examples to a header file and renamed images

tpk32 committed:
Added include and src to directory structure

tpk32 committed:
updated readme.md

tpk32 commented:
Changes made:
- created `Example.h` and `Example.cpp` for example functions
- changed directory structure to keep things clean
  - `MinGL.h` and `Example.h` header files moved to `include` folder
  - `MinGL.cpp` and `Example.cpp`  files moved to `src` folder
- changed `Main.cpp` to use example functions from `Example.h` header
- updated `readme.md`
  - changed look of directory structure
  - added compiling instructions

amengol reviewed (changes requested):
I appreciate what you did, really I do, but this is a library and you restructured it like a runtime C++ project.
In other words, I think MinGL.h and MinGL.cpp (and other core files) are better off in the main level as before. However, I am fine with an "Example" folder with everything example related in it. Sound good?

tpk32 commented:
Sure, I'll make the changes today. I am really loving working on MinGL. Also helps me learn a lot from your experience. Haha..

tpk32 committed:
restored old directory structure

tpk32 commented:
Changes made:
- Moved `MinGL.cpp` and `MinGL.h` to main level as before
- Moved `Example.cpp` and `Example.h` to Example folder
- Updated compiling instructions in readme.md

tpk32 committed:
updated image path in readme.md

amengol commented on README.md:
```diff
@@ -9,27 +9,62 @@ Note: Since ray-tracing is very demanding, you better of writing an image to dis
+│   └── glfw3native.h
+├── MinGL.h
+├── MinGL.cpp
+└── Main.cpp
```
Main.cpp is in the Example folder. Please correct this README to reflect it.

tpk32 replied:
Done, please tell me if any other change is required.

amengol reviewed (changes requested):
This is almost perfect. Just a README mistake and that's it. Thank you so much.

tpk32 committed:
fixed Main.cpp directory in readme.md

tpk32 committed:
corrected readme Main.cpp arrowhead

amengol reviewed (approved):
That looks very good. Thank you so much!
